Retry on httpx.exceptions

Abdelgha-4 opened the pull request:
Closes #5 

This is a simple draft on how to adresse the issue, please feel free to mold it into a more merge-worthy version as I don't have the time to further refine this (adding tests, think more on the design, etc...).

Thank you.

Abdelgha-4 commented:
Side question, is there any reason why `Retry.increment()` is not implemented in-place? i.e `self.attempts_made += 1`.
Creating a new object on each increment is computationally sub optimal, but I think there was some reasoning behind it, thread-safety maybe?

will-ockmore commented on httpx_retries/retry.py:
```diff
@@ -96,11 +103,18 @@ def __init__(
 
-        self.allowed_methods = frozenset(
-            HTTPMethod(method.upper()) for method in (allowed_methods or self.RETRYABLE_METHODS)
+        self.allowed_methods = (
+            self.RETRYABLE_METHODS
+            if allowed_methods is None
+            else frozenset(HTTPMethod(method.upper()) for method in allowed_methods)
+        )
+        self.status_forcelist = (
+            self.RETRYABLE_STATUS_CODES
+            if status_forcelist is None
+            else frozenset(HTTPStatus(int(code)) for code in status_forcelist)
```
these changes aren't in scope for this PR - let's revert

will-ockmore commented on httpx_retries/retry.py:
```diff
@@ -65,12 +65,19 @@ class Retry:
             HTTPStatus.GATEWAY_TIMEOUT,
         ]
     )
+    RETRYABLE_EXCEPTIONS: Final[Tuple[Type[httpx.HTTPError], ...]] = (
+        httpx.ConnectTimeout,
+        httpx.NetworkError,
+        httpx.ProtocolError,
+        httpx.ReadTimeout,
+    )
```
Having a look at https://www.python-httpx.org/exceptions/ , I'd suggest a different set of default exceptions:

- `httpx.TimeoutException`: all subclasses can be safely retried, so it's not necessary to specify `httpx.ConnectTimeout` and `httpx.ReadTimeout` individually.
- `httpx.RemoteProtocolError` should be used instead of `httpx.ProtocolError`, as `httpx.LocalProtocolError` can be caused by user error; in this case retries will always fail.

```suggestion
    RETRYABLE_EXCEPTIONS: Final[Tuple[Type[httpx.HTTPError], ...]] = (
        httpx.TimeoutException,
        httpx.NetworkError,
        httpx.RemoteProtocolError,
    )
```

will-ockmore commented on httpx_retries/retry.py:
```diff
@@ -199,13 +217,13 @@ def _calculate_sleep(self, headers: Union[httpx.Headers, Mapping[str, str]]) ->
         return self.backoff_strategy() if self.attempts_made > 0 else 0.0
 
-    def sleep(self, response: httpx.Response) -> None:
+    def sleep(self, response: httpx.Response | httpx.HTTPError) -> None:
```
this syntax [isn't compatible with python 3.9](https://github.com/will-ockmore/httpx-retries/actions/runs/13474874596/job/37765878408?pr=6), which we support

will-ockmore reviewed (changes requested):
This is great @Abdelgha-4 , thanks for submitting.

To get this merged, we need to fix the type errors, and add tests - I'm happy to give guidance on this as I've not yet included a contributions guide to the project.

will-ockmore commented:
To answer your side question, 

> is there any reason why `Retry.increment()` is not implemented in-place?

It's as you guessed, to simplify thread safety. It does increase memory usage, but in normal usage this should be negligible.

Abdelgha-4 commented:
@will-ockmore Thanks for the review!

I've applied the suggested changes, fixed typing and added a test for `is_retryable_exception`.
I'm not sure how to test the transport part however, as a starting point, I think this covers the unretryable_exception part:
```python
def test_unretryable_exception(mock_responses: MockResponse) -> None:
    mock_sleep, _ = mock_responses
    transport = RetryTransport()

    with patch("httpx.HTTPTransport.handle_request", side_effect=httpx.ProxyError("Proxy error")):
        with httpx.Client(transport=transport) as client:
            with pytest.raises(httpx.ProxyError, match="Proxy error"):
                client.get("https://example.com")

    assert mock_sleep.call_count == 0


@pytest.mark.asyncio
async def test_async_unretryable_exception(mock_async_responses: AsyncMockResponse) -> None:
    mock_asleep, _ = mock_async_responses
    transport = RetryTransport()

    with patch("httpx.AsyncHTTPTransport.handle_async_request", side_effect=httpx.ProxyError("Proxy error")):
        async with httpx.AsyncClient(transport=transport) as client:
            with pytest.raises(httpx.ProxyError, match="Proxy error"):
                await client.get("https://example.com")

    assert mock_asleep.call_count == 0
```

will-ockmore commented:
Those suggested tests look good. If you add additional tests to validate the behaviour of the default retryable exceptions, as well as user provided exceptions (this can be a unit test for the Retry class or a full test for the transport, either is fine), then we should be in a good place.

eg. for the default retryable exceptions:

```python
def test_retryable_exception(mock_responses: MockResponse) -> None:
    mock_sleep, _ = mock_responses
    transport = RetryTransport()

    with patch("httpx.HTTPTransport.handle_request", side_effect=httpx.ReadTimeout("Timeout!")):
        with httpx.Client(transport=transport) as client:
            client.get("https://example.com")

    assert mock_sleep.call_count == 10

@pytest.mark.asyncio
async def test_async_retryable_exception(mock_async_responses: AsyncMockResponse) -> None:
    mock_asleep, _ = mock_async_responses
    transport = RetryTransport()

    with patch("httpx.AsyncHTTPTransport.handle_async_request", side_effect=httpx.ReadTimeout("Timeout!")):
        async with httpx.AsyncClient(transport=transport) as client:
            await client.get("https://example.com")

    assert mock_asleep.call_count == 10
```

Abdelgha-4 commented:
@will-ockmore Thanks for the guidance, I've added the tests which proven useful very quickly by detecting two bugs.
I fixed the bugs and the all the tests are green now :).

will-ockmore commented:
Great stuff! The functionality looks solid.

All that's left is to update the documentation - add `retry_on_exceptions` to the Args section of the docstring, with a description of what it does. Then we're good to get this in 👏

Abdelgha-4 commented:
Thanks @will-ockmore!
I missed that! I've just added it with a small description.

will-ockmore reviewed (approved):
Great work here @Abdelgha-4
